lib/news_scraper/trainer/uri_trainer.rb

```diff
@@ -0,0 +1,62 @@
+module NewsScraper
+  module Trainer
+    module UriTrainer
+      extend self
+
+      def train(uri)
+        raw_html = Extractors::Article.new(uri: uri).extract
+        transformer = Transformers::Article.new(uri: uri, payload: raw_html)
+
+        begin
+          transformer.transform
+        rescue NewsScraper::Transformers::ScrapePatternNotDefined => e
+          NewsScraper::CLI.put_header
+          NewsScraper::CLI.log("There is no scrape pattern defined for #{e.root_domain} in config/article_scrape_patterns.yml")
+          no_scrape_defined(uri, raw_html, e.root_domain)
+        end
+      end
+
+      def no_scrape_defined(uri, raw_html, root_domain)
+        if NewsScraper::CLI.confirm("Step through presets for #{root_domain}?")
+          NewsScraper::CLI.put_footer
+          all_presets = current_presets
+          selected_presets = NewsScraper::Trainer::DataType.train(uri, raw_html, all_presets)
+          NewsScraper::CLI.log(selected_presets.to_yaml)
+
+          if NewsScraper::CLI.confirm("Save these scrape patterns for #{root_domain}?")
+            save_selected_presets(root_domain, selected_presets)
+          end
+        else
+          NewsScraper::CLI.log("Ignoring step-through for #{root_domain}", color: '\x1b[31m')
+          NewsScraper::CLI.put_footer
+        end
+      end
+
+      def save_selected_presets(root_domain, selected_presets)
+        save_domain_presets = if article_scrape_patterns['domains'].key?(root_domain)
+          NewsScraper::CLI.log("YAML file already contains scrape pattern for #{root_domain}:")
+          NewsScraper::CLI.log(article_scrape_patterns['domains'][root_domain].to_yaml)
+          NewsScraper::CLI.confirm("Overwrite?")
+        else
+          true
+        end
+
+        if save_domain_presets
+          article_scrape_patterns['domains'][root_domain] = selected_presets
+          File.write('config/article_scrape_patterns.yml', article_scrape_patterns.to_yaml)
+          NewsScraper::CLI.log("Successfully wrote presets for #{root_domain} to config/article_scrape_patterns.yml.")
+        else
+          NewsScraper::CLI.log("Did not write presets for #{root_domain} to file.")
+        end
+      end
+
+      def current_presets
+        article_scrape_patterns['presets']
+      end
+
+      def article_scrape_patterns
+        YAML.load_file('config/article_scrape_patterns.yml')
```

Comment:
How about

```
def article_scrape_patterns(reload: false)
   if reload || @article_scrape_patterns.nil?
      @article_scrape_patterns = YAML.load_file('config/article_scrape_patterns.yml')
   else
      @article_scrape_patterns
   end
end
```

Then we can remove `def current_presets` (it's just a key lookup on top of `article_scrape_patterns` and replace https://github.com/richardwu/news_scraper/pull/1/files/d88c2228bc676b3b9f56f547d5f9b7a4b2eedf05#diff-1ee27f6b90d971e3cdcf966d679cfb11R22 with
```
all_presets = article_scrape_patterns(reload: true)['presets']
```

That we can selectively reload (at the beginning of each `transform`) but cache it for subsequent calls in the same `tranform`.